File: client/src/services/exchange-rate.ts
import { api } from "@/lib/api";

export interface RateConversionResult {
  fromCurrency: string;
  toCurrency: string;
  amount: string;
  rate: string;
  converted: string;
  date: string;
  provider?: string;
  rateType?: "bid" | "ask" | "mid";
  sourceTimestamp?: string;
  fetchedAt?: string;
  settlementEligible: boolean;
  settlementQuoteId?: string;
}

export interface LatestRatesResult {
  baseCurrency: string;
  date: string;
  provider: string;
  rateType: "bid" | "ask" | "mid";
  rates: Record<string, string>;
}

export interface SettlementQuoteRequest {
  fromCurrency: string;
  toCurrency: string;
  amount: string;
  rateType?: "bid" | "ask" | "mid";
  date?: string;
}

export interface SettlementQuote {
  id: string;
  provider: string;
  fromCurrency: string;
  toCurrency: string;
  amount: string;
  rate: string;
  convertedAmount: string;
  rateType: "bid" | "ask" | "mid";
  sourceTimestamp: string;
  fetchedAt: string;
  expiresAt: string;
}

export class ExchangeRateService {
  async convert(from: string, to: string, amount: number, date?: string) {
    const searchParams = new URLSearchParams({ from, to, amount: String(amount) });
    if (date) searchParams.set("date", date);
    return api.get<RateConversionResult>(`/exchange-rates/convert/?${searchParams}`);
  }

  async latest(base: string) {
    const searchParams = new URLSearchParams({ base });
    return api.get<LatestRatesResult>(`/exchange-rates/latest/?${searchParams}`);
  }

  async refresh(base: string) {
    const searchParams = new URLSearchParams({ base });
    return api.post<{ status: string }>(`/exchange-rates/refresh/?${searchParams}`);
  }

  async createSettlementQuote(payload: SettlementQuoteRequest) {
    return api.post<SettlementQuote>("/exchange-rates/settlement-quotes/", payload);
  }
}

File: services/tms/internal/core/ports/services/exchangerate.go
package services

import (
	"context"
	"time"

	"github.com/emoss08/trenova/internal/core/domain/exchangerate"
	"github.com/emoss08/trenova/pkg/pagination"
	"github.com/shopspring/decimal"
)

type RateConversionResult struct {
	FromCurrency       string          `json:"fromCurrency"`
	ToCurrency         string          `json:"toCurrency"`
	Amount             decimal.Decimal `json:"amount"`
	Rate               decimal.Decimal `json:"rate"`
	Converted          decimal.Decimal `json:"converted"`
	Date               string          `json:"date"`
	Provider           string          `json:"provider,omitempty"`
	RateType           string          `json:"rateType,omitempty"`
	SourceTimestamp    *time.Time      `json:"sourceTimestamp,omitempty"`
	FetchedAt          *time.Time      `json:"fetchedAt,omitempty"`
	SettlementEligible bool            `json:"settlementEligible"`
	SettlementQuoteID  string          `json:"settlementQuoteId,omitempty"`
}

type LatestRatesResult struct {
	BaseCurrency string                     `json:"baseCurrency"`
	Date         string                     `json:"date"`
	Provider     string                     `json:"provider"`
	RateType     string                     `json:"rateType"`
	Rates        map[string]decimal.Decimal `json:"rates"`
}

type CreateSettlementQuoteRequest struct {
	FromCurrency string                `json:"fromCurrency"`
	ToCurrency   string                `json:"toCurrency"`
	Amount       decimal.Decimal       `json:"amount"`
	RateType     exchangerate.RateType `json:"rateType,omitempty"`
	Date         string                `json:"date,omitempty"`
}

type ExchangeRateService interface {
	Convert(
		ctx context.Context,
		tenantInfo pagination.TenantInfo,
		fromCurrency, toCurrency string,
		amount decimal.Decimal,
		date time.Time,
	) (*RateConversionResult, error)
	GetLatestRates(
		ctx context.Context,
		tenantInfo pagination.TenantInfo,
		baseCurrency string,
	) (*LatestRatesResult, error)
	CreateSettlementQuote(
		ctx context.Context,
		tenantInfo pagination.TenantInfo,
		req *CreateSettlementQuoteRequest,
	) (*exchangerate.SettlementQuote, error)
	RefreshRates(ctx context.Context, tenantInfo pagination.TenantInfo, baseCurrency string) error
}
